include/nuttx/semaphore.h

```diff
@@ -59,6 +60,50 @@
      {(c), (f), SEM_WAITLIST_INITIALIZER}
 #endif /* CONFIG_PRIORITY_INHERITANCE */
 
+/* Fast mutex lock/unlock path */
+
+#if defined(CONFIG_PRIORITY_PROTECT) || defined(CONFIG_PRIORITY_INHERITANCE)
+#define NXSEM_IF_FASTMUTEX(sem)                                   \
+  (((sem)->flags & (SEM_TYPE_MUTEX | SEM_PRIO_MASK)) ==           \
+   (SEM_TYPE_MUTEX | SEM_PRIO_NONE))
+#else
+#define NXSEM_IF_FASTMUTEX(sem) ((sem)->flags & SEM_TYPE_MUTEX)
+#endif
+
+/* Try wait; sets ret to OK or -EAGAIN when success,
+ * -EPERM when fast locking can't be performed
+ */
+
+#define NXSEM_FAST_TRYWAIT_IF_MUTEX(sem, ret)                           \
+  {                                                                     \
+    ret = -EPERM;                                                       \
+    if (NXSEM_IF_FASTMUTEX(sem))                                        \
+      {                                                                 \
+        int old = 1;                                                    \
+        ret = atomic_try_cmpxchg_acquire(NXSEM_COUNT(sem), &old, 0) ?   \
```

Comment:
The atomic functions cannot be used in userspace if the compiler / chip does not support atomic operations. Because in that case these would fall back to the nuttx atomic library which uses spinlocks and that is a no-no in userspace.